File: python/day03.py
import dataclasses
import io
from collections.abc import Callable, Generator

from utils import read


def peek(buf: io.TextIOWrapper) -> str:
    pos = buf.tell()
    c = buf.read(1)
    buf.seek(pos)
    return c


def readable(buf: io.TextIOWrapper) -> bool:
    return peek(buf) != ""


def read_while(
    buf: io.TextIOWrapper,
    pred: Callable[[str], bool],
) -> Generator[str]:
    while (c := peek(buf)) != "":
        if not pred(c):
            break
        buf.read(1)
        yield c


def expect(buf: io.TextIOWrapper, s: str) -> bool:
    pos = buf.tell()
    if buf.read(len(s)) == s:
        return True
    buf.seek(pos)
    return False


def read_int(buf: io.TextIOWrapper) -> int | None:
    num = "".join(read_while(buf, lambda x: x.isdigit()))
    return int(num) if num else None


@dataclasses.dataclass
class Do: ...


@dataclasses.dataclass
class Dont: ...


@dataclasses.dataclass
class Mul:
    lhs: int
    rhs: int


def read_mul(buf: io.TextIOWrapper) -> Mul | None:
    # consume
    for _ in read_while(buf, lambda b: b != "m"):
        continue

    if (
        expect(buf, "mul(")
        and ((lhs := read_int(buf)) is not None)
        and expect(buf, ",")
        and ((rhs := read_int(buf)) is not None)
        and expect(buf, ")")
    ):
        return Mul(lhs, rhs)

    # skip 1 to avoid infinite loop
    buf.read(1)
    return None


def read_all(buf: io.TextIOWrapper) -> Do | Dont | Mul | None:
    # consume
    for _ in read_while(buf, lambda b: b not in ["m", "d"]):
        continue

    if expect(buf, "do()"):
        return Do()

    if expect(buf, "don't()"):
        return Dont()

    return read_mul(buf)


def read_instructions() -> list[Do | Dont | Mul]:
    instructions = []
    with read(day=3, sample=False) as buf:
        while readable(buf):
            if (ins := read_all(buf)) is not None:
                instructions.append(ins)
    return instructions


def part1():
    ans = 0
    instructions = read_instructions()

    for ins in instructions:
        if isinstance(ins, Mul):
            ans += ins.lhs * ins.rhs
    print(ans)


def part2():
    ans = 0
    instructions = read_instructions()

    latest = Do()
    for ins in instructions:
        if isinstance(ins, Mul) and isinstance(latest, Do):
            ans += ins.lhs * ins.rhs
        elif not isinstance(ins, Mul):
            latest = ins

    print(ans)


if __name__ == "__main__":
    part1()
    part2()

File: rust/src/bin/day03.rs
use std::{
    fs::File,
    io::{BufReader, Read, Seek},
};

use anyhow::Result;

use anyhow::bail;
use rust::read;

#[derive(Debug)]
enum Instruction {
    Do,
    Dont,
    Invalid,
    Mul { lhs: i32, rhs: i32 },
}

fn read_one(reader: &mut BufReader<File>) -> Result<u8> {
    let mut buf = [0];
    reader.read_exact(&mut buf)?;
    Ok(buf[0])
}

fn read_while(reader: &mut BufReader<File>, pred: fn(&u8) -> bool) -> Vec<u8> {
    let mut out = vec![];

    while let Ok(c) = read_one(reader) {
        if !pred(&c) {
            reader.seek_relative(-1).expect("This must pass");
            break;
        }
        out.push(c);
    }

    out
}

fn expect_string(reader: &mut BufReader<File>, s: &str) -> anyhow::Result<()> {
    let pos = reader.stream_position()?;

    let expected = s.as_bytes();
    let mut buf = vec![0u8; expected.len()];

    reader.read_exact(&mut buf)?;

    if buf != expected {
        reader.seek(std::io::SeekFrom::Start(pos))?;
        bail!("Unexpected")
    }

    Ok(())
}

fn read_mul_instruction(reader: &mut BufReader<File>) -> Result<Instruction> {
    read_while(reader, |b| b != &b'm');
    expect_string(reader, "mul(")?;

    let lhs = read_while(reader, |b| (&b'0'..=&b'9').contains(&b));
    let lhs = String::from_utf8(lhs)?.parse::<i32>()?;

    expect_string(reader, ",")?;

    let rhs = read_while(reader, |b| (&b'0'..=&b'9').contains(&b));
    let rhs = String::from_utf8(rhs)?.parse::<i32>()?;

    expect_string(reader, ")")?;

    Ok(Instruction::Mul { lhs, rhs })
}

fn read_instruction(reader: &mut BufReader<File>) -> Result<Instruction> {
    read_while(reader, |b| b != &b'm' && b != &b'd');

    if expect_string(reader, "do()").is_ok() {
        return Ok(Instruction::Do);
    }
    if expect_string(reader, "don't()").is_ok() {
        return Ok(Instruction::Dont);
    }

    match read_mul_instruction(reader) {
        Ok(mul) => Ok(mul),
        Err(_) => {
            read_one(reader)?;
            Ok(Instruction::Invalid)
        }
    }
}

fn read_instructions() -> Result<Vec<Instruction>> {
    let mut reader = read(3, false)?;

    let mut instructions = vec![];
    while let Ok(ins) = read_instruction(&mut reader) {
        instructions.push(ins)
    }
    Ok(instructions)
}

fn part1() -> Result<()> {
    let instructions = read_instructions()?;
    let sum = instructions
        .into_iter()
        .map(|ins| match ins {
            Instruction::Mul { lhs, rhs } => lhs * rhs,
            _ => 0,
        })
        .sum::<i32>();
    println!("{}", sum);
    Ok(())
}

fn part2() -> Result<()> {
    let instructions = read_instructions()?;
    let mut sum = 0;

    let mut last = Instruction::Do;
    for ins in instructions {
        match ins {
            ins @ (Instruction::Do | Instruction::Dont) => last = ins,
            Instruction::Mul { lhs, rhs } => {
                if matches!(last, Instruction::Do) {
                    sum += lhs * rhs
                }
            }
            _ => {}
        }
    }
    println!("{}", sum);
    Ok(())
}

fn main() -> Result<()> {
    part1()?;
    part2()?;
    Ok(())
}
